File: tools/e2e-connect-web/src/flows/concurrency.test.ts
import { describe, it, expect } from "vitest";
import { services, clientFor, range } from "../harness.js";

/**
 * Many in-flight unary calls on a single client must not cross-talk: every
 * response has to match its own request, proving the dispatcher keeps concurrent
 * calls independent.
 */
const CONCURRENT_CALLS = 64;

for (const service of services) {
  describe(`${service.name} · concurrency`, () => {
    const client = clientFor(service.schema);

    it(`keeps ${CONCURRENT_CALLS} concurrent unary calls independent`, async () => {
      const indices = range(0, CONCURRENT_CALLS - 1);

      const responses = await Promise.all(indices.map((index) => client.echo({ message: `c${index}` })));

      expect(responses.map((response) => response.message)).toEqual(indices.map((index) => `echo: c${index}`));
    });
  });
}

File: project/app-server-spring/src/test/kotlin/io/github/cgardev/example/protocol/ConcurrencyTest.kt
package io.github.cgardev.example.protocol

import io.github.cgardev.example.v1.EchoRequest
import io.github.cgardev.example.v1.EchoResponse
import org.junit.jupiter.api.Assertions.assertEquals
import java.util.concurrent.Callable
import java.util.concurrent.Executors

/**
 * Many in-flight unary requests must not cross-talk: every response has to match
 * its own request, proving the dispatcher keeps concurrent calls independent.
 */
class ConcurrencyTest : ProtocolTestSupport() {

    @BothServices
    fun `handles many concurrent requests without cross-talk`(service: String) {
        val count = 64
        val pool = Executors.newFixedThreadPool(16)
        try {
            val tasks = (0 until count).map { index ->
                Callable {
                    val body = EchoRequest.newBuilder().setMessage("c$index").build().toByteArray()
                    val response = post(service, "Echo", "application/proto", body)
                    assertEquals(200, response.statusCode())
                    EchoResponse.parseFrom(response.body()).message
                }
            }
            val results = pool.invokeAll(tasks).map { it.get() }
            assertEquals((0 until count).map { "echo: c$it" }.toSet(), results.toSet())
        } finally {
            pool.shutdownNow()
        }
    }
}
